// Copyright 2022 The XLS Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto2";

package xls.verilog;

// A span of lines in a source code (DSLX or C++ or Verilog) file.
// This may later be augmented to support column numbers.
message SourceSpan {
  optional int64 line_start = 1;
  optional int64 line_end = 2;
}

// A mapping between a span in a source file (DSLX or C++ in the case of xlscc)
// and a span in a generated Verilog file.
message VerilogLineMapping {
  optional string source_file = 1;
  optional SourceSpan source_span = 2;
  optional string verilog_file = 3;
  optional SourceSpan verilog_span = 4;
}

// A relation between spans in input source files and spans in generated Verilog
// files. This relation is not necessarily a function -- multiple source spans
// can correspond to one Verilog span, and multiple Verilog spans can correspond
// to one source span.
//
// The intended semantics is that a row in this relation witnesses a source code
// element (spanning some lines) that gives rise to some Verilog code (spanning
// some lines).
message VerilogLineMap {
  repeated VerilogLineMapping mapping = 1;
}
